### PR TITLE
refactor(components): refactor progress

### PR DESCRIPTION
**Pull request by Alanscut**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.
#3113

### CODE REVIEW

File: packages/components/progress/src/progress.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  percentage: {
+    type: Number,
+    default: 0,
+    validator: (val: number): boolean => val >= 0 && val <= 100,
+  },
+  status: {
+    type: String,
+    default: '',
+    validator: (val: string): boolean =>
+      ['', 'success', 'exception', 'warning'].indexOf(val) > -1,
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:27]: please use `values` property

File: packages/components/progress/src/progress.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import {
+  buildProp,
+  buildProps,
+  definePropType,
+} from '@element-plus/utils/props'
+import type { SVGAttributes } from 'vue'
+import type { ExtractPropTypes } from '@vue/runtime-core'
+
+type ProgressFuncType = (percentage: number) => string
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:9]: Suggestion: rename it to `ProgressFn`

File: packages/components/progress/src/progress.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+export const progressProps = buildProps({
+  type: {
+    type: String,
+    default: 'line',
+    validator: (val: string): boolean =>
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:15]: please use `values` property

File: packages/components/progress/src/progress.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    string | { color: string; percentage: number } | ProgressFuncType
+  >({
+    type: [String, Array, Function],
+    default: '',
+  }),
+  format: {
+    type: definePropType<ProgressFuncType>(Function),
+    default: (percentage: number): string => `${percentage}%`,
+  },
+})
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:67]: add `as const`

File: packages/components/progress/src/progress.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    type: Number,
+    default: 3,
+  },
+  strokeWidth: {
+    type: Number,
+    default: 6,
+  },
+  strokeLinecap: {
+    type: definePropType<NonNullable<SVGAttributes['stroke-linecap']>>(String),
+    default: () => 'round' as NonNullable<SVGAttributes['stroke-linecap']>,
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:43]:
```suggestion
    default: 'round',
```

File: packages/components/progress/src/progress.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import {
+  buildProp,
+  buildProps,
+  definePropType,
+} from '@element-plus/utils/props'
+import type { SVGAttributes } from 'vue'
+import type { ExtractPropTypes } from '@vue/runtime-core'
```
Comment (sxzz) [inline packages/components/progress/src/progress.ts:7]:
```suggestion
import type { ExtractPropTypes } from 'vue'
```

**Review by sxzz — CHANGES_REQUESTED** [review]
Also, please fix all typescript errors.

**Comment by Alanscut** [comment]
@sxzz The problems you raised have been fixed, please check again

**Review by sxzz — APPROVED** [review]

**Pull request merged**